File: loom/loom-common/src/main/java/com/hp/hpl/loom/model/QueryResult.java
/*******************************************************************************
 * (c) Copyright 2017 Hewlett Packard Enterprise Development LP Licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in compliance with the License. You
 * may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *******************************************************************************/
package com.hp.hpl.loom.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

/**
 * The result of a query operation to retrieve a Thread. An aggregation of ThreadResultElements.
 * Used for on-the-wire wrapping.
 */
@JsonAutoDetect
public class QueryResult {

    /**
     * The status for the QueryResult.
     */
    public enum Statuses {
        /**
         * QueryResult READY status.
         */
        READY,
        /**
         * QueryResult PENDING status.
         */
        PENDING
    };

    /**
     * Logical ID of the DA for the result of the query.
     */
    private String logicalId;

    /**
     * Optional name for the result.
     */
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String name;

    private ArrayList<QueryResultElement> elements;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private ItemType itemType;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Statuses status;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Long alertCount;

    /**
     * Highest alert level in the cluster.
     */
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Integer highestAlertLevel;

    /**
     * Description of (one of the) highest alert level(s) in the cluster.
     */
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String highestAlertDescription;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)

    private Map<String, Long> timings = new HashMap<>();
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Long totalTime;
    private QueryResultElement excludedItems;

    /**
     * Default constructor.
     */
    public QueryResult() {
        elements = new ArrayList<QueryResultElement>();
        status = Statuses.READY;
    }

    /**
     * Constructor with logicalId.
     *
     * @param logicalId the logicalId
     */
    public QueryResult(final String logicalId) {
        super();
        this.logicalId = logicalId;
    }

    /**
     * Constructor with a logicalId and name.
     *
     * @param logicalId the logicalId
     * @param name the name
     */
    public QueryResult(final String logicalId, final String name) {
        super();
        this.logicalId = logicalId;
        this.name = name;
    }

    /**
     * Constructor with a aggregation.
     *
     * @param aggregation aggregation to construct from
     */
    public QueryResult(final Aggregation aggregation) {
        elements = new ArrayList<QueryResultElement>(aggregation.getElements().size());
        logicalId = aggregation.getLogicalId();
        for (Fibre entity : aggregation.getElements()) {
            QueryResultElement queryResultElement = new QueryResultElement(entity);
            addElement(queryResultElement);
        }
        status = Statuses.READY;

        alertCount = aggregation.getAlertCount();
        highestAlertDescription = aggregation.getHighestAlertDescription();
        highestAlertLevel = aggregation.getHighestAlertLevel();
    }

    /**
     * Get the alertCount value.
     *
     * @return the alertCount
     */
    public Long getAlertCount() {
        return alertCount;
    }

    /**
     * Set the alertCount value.
     *
     * @param alertCount the alertCount
     */
    public void setAlertCount(final Long alertCount) {
        this.alertCount = alertCount;
    }

    /**
     * Get the highestAlertLevel.
     *
     * @return the highestAlertLevel
     */
    public Integer getHighestAlertLevel() {
        return highestAlertLevel;
    }

    /**
     * Set the highestAlertLevel.
     *
     * @param highestAlertLevel the highestAlertLevel
     */
    public void setHighestAlertLevel(final Integer highestAlertLevel) {
        this.highestAlertLevel = highestAlertLevel;
    }

    /**
     * Get the highestAlertDescription.
     *
     * @return the highestAlertDescription
     */
    public String getHighestAlertDescription() {
        return highestAlertDescription;
    }

    /**
     * Set the HighestAlertDescription.
     *
     * @param highestAlertDescription the highestAlertDescription
     */
    public void setHighestAlertDescription(final String highestAlertDescription) {
        this.highestAlertDescription = highestAlertDescription;
    }

    /**
     * Get the statuses.
     *
     * @return the statuses
     */
    public Statuses getStatus() {
        return status;
    }

    /**
     * Set the status.
     *
     * @param status the status
     */
    public void setStatus(final Statuses status) {
        this.status = status;
    }

    /**
     * Get the logicalId.
     *
     * @return the logicalId
     */
    public String getLogicalId() {
        return logicalId;
    }

    /**
     * Set the logicalId.
     *
     * @param logicalId the logicalId
     */
    public void setLogicalId(final String logicalId) {
        this.logicalId = logicalId;
    }

    /**
     * Get the name.
     *
     * @return the name
     */
    public String getName() {
        return name;
    }

    /**
     * Set the name.
     *
     * @param name the name
     */
    public void setName(final String name) {
        this.name = name;
    }

    /**
     * Get the query result elements.
     *
     * @return the query result elements list.
     */
    public ArrayList<QueryResultElement> getElements() {
        return elements;
    }

    /**
     * Set the query result elements.
     *
     * @param elements the query result elements list.
     */
    public void setElements(final ArrayList<QueryResultElement> elements) {
        this.elements = elements;
    }

    /**
     * Add the queryResultElement.
     *
     * @param queryResultElement the queryResultElement
     */
    public void addElement(final QueryResultElement queryResultElement) {
        elements.add(queryResultElement);
    }

    /**
     * Get the ItemType.
     *
     * @return the itemType
     */
    public ItemType getItemType() {
        return itemType;
    }

    /**
     * Set the ItemType.
     *
     * @param itemType the itemType
     */
    public void setItemType(final ItemType itemType) {
        this.itemType = itemType;
    }

    /**
     * Calculate total number of relations across all Resource elements.
     *
     * @return the total size of the relations.
     */
    public int totalRelations() {
        int total = 0;
        for (QueryResultElement resource : getElements()) {
            total += resource.getRelations().size();
        }
        return total;
    }

    @Override
    public String toString() {
        ObjectMapper mapper = new ObjectMapper();
        mapper.enable(SerializationFeature.INDENT_OUTPUT);
        String jsonRep = "";
        try {
            jsonRep = mapper.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            System.out.println("JsonProcessingException " + e);
        }
        return jsonRep;
    }

    /**
     * Recursively get all Fibre contained by Aggregation and all of its children. It returns a map
     * of the containing aggregation to the fibre.
     *
     * @return a map of the aggregation or fibre logical id to the item logical id.
     */
    @JsonIgnore
    public final HashMap<String, String> getContainedItemsLogicalIds() {
        HashMap<String, String> logicalIds = new HashMap<>();
        List<QueryResultElement> items = this.getElements();
        for (QueryResultElement queryResultElement : items) {
            Fibre fibre = queryResultElement.getEntity();
            if (fibre.isAggregation()) {
                Aggregation agg = (Aggregation) fibre;
                getItems(logicalIds, agg);
            } else {
                logicalIds.put(fibre.getLogicalId(), this.getLogicalId());
            }
        }
        return logicalIds;
    }

    /**
     * Walks the given aggregation adding in the items + containing logical id to the map provided.
     *
     * @param map The aggregation/fibre logical id to the item logical id
     * @param aggregation The aggregation to walk down
     */
    private void getItems(final HashMap<String, String> map, final Aggregation aggregation) {
        List<Fibre> fibres = aggregation.getElements();
        for (Fibre fibre : fibres) {
            if (fibre.isAggregation()) {
                this.getItems(map, (Aggregation) fibre);
            } else {
                map.put(fibre.getLogicalId(), aggregation.getLogicalId());
            }
        }
    }

    /**
     * Add timing information.
     *
     * @param action action we are timing for
     * @param time the time
     */
    public void addTiming(String action, long time) {
        timings.put(action, time);
    }

    /**
     * Set the total time.
     *
     * @param time the time
     */
    public void setTotalTiming(Long time) {
        this.totalTime = time;
    }

    /**
     * Get the totalTiming.
     *
     * @return the total timing
     */
    public Long getTotalTiming() {
        return totalTime;
    }

    /**
     * Get the timing info back for this query result.
     *
     * @return get the timing for this query result
     */
    public Map<String, Long> getTimings() {
        return timings;
    }

    /**
     * @return the excludedItems
     */
    public QueryResultElement getExcludedItems() {
        return excludedItems;
    }

    /**
     * @param excludedItems the excludedItems to set
     */
    public void setExcludedItems(QueryResultElement excludedItems) {
        this.excludedItems = excludedItems;
    }
}

File: front-end/lib/weft/js/weft/models/QueryResult.js
/*******************************************************************************
 * (c) Copyright 2017 Hewlett Packard Enterprise Development LP Licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in compliance with the License. You
 * may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *******************************************************************************/
define(function (require) {

  "use strict";

  var _ = require('lodash');
  var Backbone = require('backbone');
  var Item = require('weft/models/Item');
  var Aggregation = require('weft/models/Aggregation');
  var NewElementsMarker = require('./QueryResult/NewElementsMarker');
  var EventBus = require('weaver/utils/EventBus');

  /**
   * Representation of the result of a Query
   * @class QueryResult
   * @namespace models
   * @module weft
   * @constructor
   * @type {[type]}
   */
  var QueryResult = Backbone.Model.extend({

    urlRoot: '/results',

    /**
     * @property ENTITY_TYPE_AGGREGATION
     * @type {String}
     * @default Aggregation
     * @final
     */
    ENTITY_TYPE_AGGREGATION: 'Aggregation',

    defaults: function () {

      return {
        /**
         * The item type of the elements of the QueryResult
         * @attribute  itemType
         * @type {ItemType|AggregationItemType}
         */

        /**
         *
         * @attribute pending
         * @type {Boolean}
         * @default  true
         */
        pending: true,

        /**
         * Flag marking if the result do not correspond
         * to the query of their Thread
         * @type {Boolean}
         */
        outdated: false,

        /**
         * The elements
         * @attribute elements
         * @type {Backbone.Collection}
         */
        elements: undefined, // Makes sure an event gets fired when results come back empty

        /**
         * The list of Threads using this result
         * @attribute  threads
         * @type {Backbone.Collection}
         */
        threads: new Backbone.Collection()

        /**
         * An Aggregation representing items that were excluded from the results
         * due to some filtering happening in the query
         * @attribute excludedItems
         * @type {models.Aggregation}
         */
      };
    },

    initialize: function () {
      /**
       * @todo This should be made clearer that it provides new abilities via event listeners
       * I'm not sure about this pattern, it happens a lot in the codebase and seems to make things more complicated
       * than they need to be. The ability isn't injected in, its hard coded.. so why not have it directly on the class?
       */
      new NewElementsMarker(this);
    },

    /**
     * Checks if this QueryResult contains given element
     * @method  hasElement
     * @param  {models.Element}  element The element that might be contained in this QueryResult
     * @return {Boolean}
     */
    hasElement: function (element) {
      return this.get('elements') ? this.get('elements').contains(element) : false;
    },

    /**
     * Refreshes the result. Differs from Backbone.Model.fetch by processing the resulting elements asynchronously
     * @method  refresh
     * @param fibersIndex {Backbone.Collection}  The list of fibers into which Items and Aggregations
     *                                          should be looked up when parsing the response
     * @return {jqXHR} The jQuery XHR object used to request the results
     */
    refresh: function (fibersIndex) {
      this.cancelRefresh();
      // IMPROVE: A bit clunky to delegate back to standard Backbone `sync`
      if (this.get('aggregator')) {
        this._xhr = this.get('aggregator').send({
          type: 'GET',
          url: this.url()
        });
      } else {
        this._xhr = this.sync('read', this);
      }

      var self = this;
      this._xhr.then(function _handleRefreshResponse(response) {
        var parsedResponse = self.parse(response, fibersIndex);
        self.set(parsedResponse);
        // Refresh the sort on each threads (LOOM-1619)
        // todo: Why is it there and not at the Thread level, listening for 'reset:elements'?
        self.get('threads').forEach(function (thread) {
          thread.sortBy(thread.get('sort'));
        });

        var aggregator = self.get('aggregator');
        if (aggregator) {
          var selectedFiber = aggregator.get('selectedFiber');
          if (selectedFiber) {
            EventBus.trigger('fiber:custom:reselect', {fiberView: selectedFiber});
          }
        }

      });
      this._xhr.always(_.bind(function () {
        this._xhr = null;
      }, this));
      return this._xhr;
    },

    /**
     * @method  isRefreshing
     * @return {Boolean} `true` if this QueryResult is currently being refreshed, `false` otherwise
     */
    isRefreshing: function () {
      return !!this._xhr;
    },

    /**
     * Cancels current refresh if there is one
     * @method  cancelRefresh
     */
    cancelRefresh: function () {
      if (this._xhr) {
        this._xhr.abort();
      }
    },

    /**
     * Override Backbone's sync method to allow CORS requests
     * @param method
     * @param model
     * @param options
     * @returns {*|JQueryXHR}
     */
    sync: function (method, model, options) {
      options = options || {};
      options.crossDomain = true;
      options.xhrFields = options.xhrFields || {};
      options.xhrFields.withCredentials = true;
      return Backbone.Model.prototype.sync.call(this, method, model, options);
    },

    /**
     * Adds an index attribute to elements so they can be sorted back to their original order
     * @method  _addOriginalSortIndexAttribute
     * @param json
     * @private
     */
    _addOriginalSortIndexAttribute: function (json) {
      _.forEach(json.elements, function (element, index) {
        element.entity['l.index'] = index;
      });
    },

    /**
     * Parse the JSON result
     * @method parse
     * @param json
     * @param fibersIndex
     * @returns {{pending: boolean, contentType: (Array|*)}}
     */
    parse: function (json, fibersIndex) {
      var result = {
        pending: json.status === 'PENDING',
        contentType: json.elements && json.elements[0] && json.elements[0].entity['l.entityType']
      };
      if (json.excludedItems) {
        result.excludedItems = this._parseElement(json.excludedItems, json.itemType, fibersIndex);
      }
      this._addOriginalSortIndexAttribute(json);
      var parsedElements = this._parseElements(json, fibersIndex);
      result.elements = new Backbone.Collection(parsedElements);
      result.itemType = result.elements.at(0) ? result.elements.at(0).itemType : undefined;
      return result;
    },

    /**
     * Parse the elements from the JSON response
     * @param json
     * @param fibersIndex
     * @private
     */
    _parseElements: function (json, fibersIndex) {
      return _.map(json.elements, function (element) {
        return this._parseElement(element, json.itemType, fibersIndex);
      }, this);
    },

    /**
     * Parse an individual element from the JSON response
     * @param element
     * @param itemType
     * @param fibersIndex
     * @returns {*}
     * @private
     */
    _parseElement: function (element, itemType, fibersIndex) {
      _.merge(element.entity, _.omit(element, 'entity'));
      if (element.entity['l.entityType'] === this.ENTITY_TYPE_AGGREGATION) {
        this._parseAggregatedAttributes(element.entity);
      }
      var instance = this._getElementInstance(element.entity, itemType, fibersIndex);
      var alertProperties = this._parseAlert(element.entity);
      instance.alert.set(alertProperties);
      // let the model know we have just synced values into it
      instance.trigger('sync', element.entity);
      return instance;
    },

    /**
     * Parse alerts out of the JSON response
     * @param json
     * @returns {*}
     * @private
     */
    _parseAlert: function (json) {
      if (json['l.entityType'] === this.ENTITY_TYPE_AGGREGATION) {
        return {
          level: json.highestAlertLevel || 0,
          description: json.highestAlertDescription,
          count: json.alertCount || 0
        };
      } else {
        return {
          level: json['l.alertLevel'] || 0,
          description: json['l.alertDescription'] || '',
          count: 0
        };
      }
    },

    /**
     * This method is responsible for extracting the plottable aggregate stats from the response. It is where we
     * currently filter out everything apart from the avg value.
     *
     * It takes the matched plottable items, transforms the name, and maps the values directly onto the JSON
     * object passed. so JSON.<plottableX>.<Name>_<Variant> becomes JSON.<Name>, meaning the variant part is lost.
     * todo: All Weaver components expect this stripped named value to be directly on the object without the variant.
     *
     * @method _parseAggregatedAttributes
     * @param json
     * @private
     */
    _parseAggregatedAttributes: function (json) {
      var values = _.reduce(json.plottableAggregateStats, function (values, value, id) {
        if (id.indexOf('_avg') !== -1) {
          values[id.substr(0, id.length - 4)] = value;
        }
        return values;
      }, {});
      _.merge(json, values);
      // values written into metricsValues are consumed by the metrics history
      json.metricsValues = values;
    },

    /**
     * @method _getElementInstance
     * @param entity
     * @param itemType
     * @param fibersIndex
     * @returns {Aggregation|Item}
     * @private
     */
    _getElementInstance: function (entity, itemType, fibersIndex) {
      var currentInstance = fibersIndex.get(this._extractEntityId(entity));
      if (currentInstance) {
        currentInstance.set(entity);
        return currentInstance;
      }
      return this._getNewInstance(entity, itemType);
    },

    /**
     * Extracts the correct id for either an Aggregation or a Logical Item
     * @method _extractEntityId
     * @param entity
     * @returns {*}
     * @private
     */
    _extractEntityId: function (entity) {
      return entity['l.entityType'] === this.ENTITY_TYPE_AGGREGATION ? entity['l.semanticId'] : entity['l.logicalId'];
    },

    /**
     * @method _getNewInstance
     * @param entity
     * @param itemType
     * @returns {Aggregation|Item}
     * @private
     */
    _getNewInstance: function (entity, itemType) {
      if (entity['l.entityType'] === this.ENTITY_TYPE_AGGREGATION) {
        return new Aggregation(entity, {itemType: itemType});
      }
      return new Item(entity, {itemType: itemType});
    },

    /**
     * @method url
     * @returns {string}
     */
    url: function () {
      var thread = this.get('threads').at(0);
      var context = thread ? thread.url() : '';
      return context + this.urlRoot;
    }
  });

  return QueryResult;
});
